### PR TITLE
Replace flake8 with ruff linter

### PR DESCRIPTION
**Pull request by ItsDrike**
This replaces `flake8` in favor of [`ruff`](https://beta.ruff.rs/docs/), which is a linter written in rust, and runs much faster (10-100x speedup).

It has first party support for all of the flake8 extensions which we were using (except for `flake8-new-union-types`, but there is an equivalent rule in the `pyupgrade` rule set). It also supports configuration directly from `pyproject.toml`, meaning we lo longer need yet another config file in the project's root (`.flake8`).

Ruff is almost completely compatible with flake8, so no breaking changes in code style are introduced here.

Amazingly enough, ruff also has `--fix` argument, which can be used to automatically fix any rule violations that  it knows how to fix. This can end up greatly reducing a lot of manual labor, especially for new contributors not used to our linting rules.

---

I also enabled `RUF` ruleset, which contains various custom rules created by the ruff devs. Generally, these are pretty useful common sense rules and we currently don't violate almost anything there, except for these:
- `RUF012`, which enforces that class variables which are mutable should use the `typing.ClassVar` annotation is violated in some places. For now, I simply ignored this rule, however we can enable it later if we'd like, and fix the violations in another PR.
- `RUF100` , which is a rule about unnecessary noqa comments. All violations of this rule were fixed here. Note that it might sometimes seem like the noqa comment should be there, and it might seem like a bug that ruff doesn't find an error there. <br>I've added explanations for these as review comments, pointing out why they were removed, but the general reason is that ruff is just really smart, and understands when a rule should not apply. I marked these as resolved though, so they won't prevent a merge (as our branch protection doesn't allow unresolved conversations), so you'll need to click over my comments there to read them.

### CODE REVIEW

File: mcstatus/utils.py
Comment (ItsDrike) [inline mcstatus/utils.py:127]: Just a quick note about this, as it might seem like a bug:

Ruff is able to pick up that this method continas typing overloads, which themselves don't use the `Any` type, and therefore the use of `Any` in the implementation here is allowed. It's a feature, not a bug.

File: tests/motd/test_transformers.py
Comment (ItsDrike) [inline tests/motd/test_transformers.py:75]: Ruff doesn't enforce W503 at all, as it's actually against the PEP 8 suggested formatting rules, this is something that was only imposed by flake8 as something they considered to be a better practice. In fact W503 clashes with black, so we wouldn't want to be enforcing it anyway.

File: mcstatus/__init__.py
Comment (ItsDrike) [inline mcstatus/__init__.py:1]: Ruff is able to pick up that this is suposed to be "re-exported", as it's present in `__all__`

File: pyproject.toml
Comment (PerchunPak) [inline pyproject.toml:57]: This is really questionable. Instead of implementing the own version of flake8's API for being backwards-compatible with all plugins, they ended up rewriting every popular plugin in rust. This gives us quite outdated version of packages and may only be worse, if `ruff` will not have strong maintaining.

Although, I am okay using it now, because it can address all our needs and we are not planning to add more flake8 plugins in the future.

File: pyproject.toml
Comment (PerchunPak) [inline pyproject.toml:52]: Why can't we also replace `isort` with `ruff`?

File: pyproject.toml
Comment (PerchunPak) [inline pyproject.toml:112]:
```suggestion
ignore = [
    "E203",   # Incompatible with black in some moments
```

See https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html#flake8

File: pyproject.toml
Comment (PerchunPak) [inline pyproject.toml:129]: Backport `ban-relative-imports=true` from `.flake8`.

```suggestion

[tool.ruff.flake8-tidy-imports]
ban-relative-imports = "all"
```

**Review by PerchunPak — CHANGES_REQUESTED** [review]
Never thought, that one day I will replace `flake8`

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:112]: This is disabled by default in ruff (part of nursery), see:
image
image
image

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:129]: oh, good catch, I intended to do this but forgot after removing .flake8

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:52]: We could, but I think it's a better idea to keep them separate for now, ruff's isort doesn't have complete feature parity with isort in terms of settings, I don't know if we use some of those unsupported settings, but I think we can hold this off for now

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:57]: It is true that Ruff could use a plugin system, and it is [already in discussion](https://github.com/astral-sh/ruff/issues/283).

However, the main point of ruff is to be really fast, to do that, first party support was preferred, especially when you consider that ruff is a rust library (it's unclear whether the plugins to ruff would then be written in python, or rust, if in python, that would likely greatly slow it down, if in rust, it might be pretty annoying to write for people not too familiar). 

IMO ruff did a really good job at re-implementing essentially all of the popular flake8 extensions, or at least providing equivalent or near equivalent (and often better)  alternatives to them form different rule-sets. I have been working on migrating various other projects of mine that use much more flake8 extensions than we do in mcstatus, and so far, ruff had support, or alternatives for all of them.

That said, there are of course a lot of much less commonly used flake8 extensions, or even completely custom ones that enforce a very specific thing unique to a codebase, and for that, ruff might not currently be the best option (at least until the plugin system does get added). However as you said, we don't really need anything too custom in mcstatus, and with everything Ruff currently supports, it should be more than enough.

---

You did also mention a point on backwards compatibility, and how Ruff will have to actively keep track of the plugins that it "carried over". I'd like to talk a bit on that, as it's pretty important to explain how Ruff handles this.

Generally, the burden of maintaining any of the ported rules come to Ruff's maintainers, and they do accept this, and aim to mostly remain compatible. 

That said, Ruff is already not 100% compatible with the ported extensions, and that has various reasons. The biggest one is that many of the ported extensions had some overlap in the rules, and so Ruff only kept one of them, sometimes modifying that one to merge the behaviors from both extensions, when that made sense. 

The other reason is that Ruff addresses the bugs reported to them directly, so in many cases, the ported rules were changed to fix a bug, or even introduce a new subfeature for that rule. That means even though a rule might have the same error code, it's not always the same rule in what it's enforcing.

This is because Ruff is not trying to be a mere 1:1 replica of flake8 and the extension that were carried over. It's very important to understand that Ruff is it's own thing, and the rules, included those originally migrated from other flake8 extensions are treated as first party rules, with their own, potentially different implementation from the one in the ported extension.

Ruff might even eventually add new rules under an existing ruleset of some extension, if it makes contextual sense, say there's a new PEP about annotations, ruff could absolutely implement a rule for it under the `flake8-annotations` rule-set, even though the original extension didn't implement it.

They generally try not to do this, to avoid any potential rule clashes if the extension later does implement it, but it's not out of question, and especially if ruff at some point becomes more popular than flake8, which it is on a pretty good track for, it's the flake8 extensions that might want to start confirming to ruff's rules, rather than the other way around. Though right now, new custom rules are being introduced purely under the `RUF` category, but they are mostly pretty general rules, applicable everywhere, and not something that could potentially belong udner any of the extensions rulesets.

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:129]: Fixed in a [force push](https://github.com/py-mine/mcstatus/compare/75287f60b0ccbc8046eeec1b181972820df64b61..1c5cc62e09c2461608cd8c1572248a561cf80459)

File: tests/status_response/__init__.py
Comment (ItsDrike) [reply tests/status_response/__init__.py:19]: Hm, apparently not, `flake8-annotations` doesn't seem to be capable of detecting `Any` in nested  types, so say `list[Any]` would not be a violation of ANN401. The docs mention this: https://github.com/sco1/flake8-annotations#dynamic-typing-caveats

Additionally, variables (including class variables) aren't enforced to not contain `Any` at all, `flake8-annotations` only checks `Any` in functions (returns or arguments).

Ruff replicates this behavior and also doesn't consider `Any` in nested types to be an issue.

---

However I have also noticed that we actually don't seem to even have `ANN401` enabled. This is an optionally enforced rule, and needs to be selected (`flake8 --extend-select=ANN401`), but if you take a look at the `.flake8` config file (now removed) didn't contain `extend-select=ANN401`, so apparently we weren't even enforcing this in the first place.

I suppose the noqas were added manually, because of the use of Any was suspected to be an issue, without actually checking? I'm not quite sure. However I did put together this minimal example, to show how `flake8-annotations` behaves on various things, and indeed it only reports issues when `Any` is used directly, and only as a function return/argument annotation:

<details>
<summary>Example code</summary>

```python
from typing import Any, ClassVar, TypeVar

T = TypeVar("T")

a: Any = 5
b: list[Any] = [1, "a", 3.5]


class Foo:
    X: ClassVar[Any] = 10
    Y: Any = 50


def func1(x: list[Any]) -> object:
    ...


def func2(x: Any) -> object:
    ...


def func3(x: object) -> Any:
    ...


def func4(x: object) -> list[Any]:
    ...
```

With just flake8 and flake8-annotations installed, running `flake8 --extend-select=ANN401 --show-source main.py` shows this output:

```
main.py:18:11: ANN401 Dynamically typed expressions (typing.Any) are disallowed
def func2(x: Any) -> object:
          ^
main.py:22:28: ANN401 Dynamically typed expressions (typing.Any) are disallowed
def func3(x: object) -> Any:
                           ^
```

</details>

File: pyproject.toml
Comment (kevinkjt2000) [reply pyproject.toml:49]: I assume that is the intent behind having the ^ here. Most of the other dev tools have this specification in pyproject.toml.

**Review by kevinkjt2000 — APPROVED** [review]

File: pyproject.toml
Comment (PerchunPak) [reply pyproject.toml:49]: > All it's likely to do in the updates is adding new rules / fix bugs on existing rues.

Exactly this can fail our CI, the same way as pyright sometimes fails after update.

> for which we also use a carrot dependency

We actually shouldn't.

> thing is, these tools can't just count every change to a rule as a breaking change, or their major version would be really high.

Mypy somehow have their [sem-ver like versioning](https://mypy-lang.blogspot.com/2023/02/mypy-10-released.html), that resolves this problem.

Comment (kevinkjt2000) [reply pyproject.toml:49]: I'm okay with updates potentially failing in CI. They are handled individually by Dependabot and allow us to preview upcoming changes. Those results enable us to report bugs if necessary. Or, by waiting until the next update PR, the problem may disappear without any intervention.

Comment (PerchunPak) [reply pyproject.toml:49]: Okay, never mind, I don't really want to block this PR because of one symbol removing.

**Review by PerchunPak — APPROVED** [review]

**Comment by ItsDrike** [comment]
Just a quick note that ruff's handling of ANN401 still does contain a bug, see: https://github.com/astral-sh/ruff/issues/5803 so we might want to hold this off until that's resolved, or disable the enforcement of this rule entirely, as it wasn't enforced by `flake8-annotations` before at all.

It is however a pretty minor bug, only concerning overloads, otherwise the implementation works exactly the same as it would with `flake8-annotations`. 

Another note I'd mention on ruff's handling of ANN401 is that as I said in one of the requested changes, ruff doesn't consider the use of Any for variables an issue, and the use of Any in generic variables is also not an issue. This is the same behavior as `flake8-annotations`, but it is potentially a bit unexpected. There is a feature request for this already: https://github.com/astral-sh/ruff/issues/5871 and there will eventually likely be new rules added to address these too.

File: pyproject.toml
Comment (ItsDrike) [reply pyproject.toml:49]: This can always be done later, in another PR which perhaps does the same for pyright and any other deps that need it. I'm still not sure we should be doing this though, so if you will make another PR for it, do add a good description explaining why it should be done.

**Comment by PerchunPak** [comment]
I don't think this is a big problem for us, so we can just ignore it for now.

File: pyproject.toml
Comment (kevinkjt2000) [reply pyproject.toml:49]: `poetry update` is a tiny part of the development workflow, and Dependabot mainly handles it. With the locked version already passing CI tests, contributors are guaranteed to have a working version unless they do Dependabot's job. If a contributor messes with package versions, some code likely needs to be updated in mcstatus or a bug in a package they are investigating.

**Pull request merged**
